netspeed/src/netspeed.c

```diff
@@ -1371,54 +1364,44 @@ update_tooltip (NetspeedApplet *netspeed)
-                if (strlen (ipv6_text) > 2) {
-                    g_string_printf (tooltip,
-                                     _("%s: %s and %s"),
-                                     netspeed->devinfo->name,
-                                     ipv4_text,
-                                     ipv6_text);
-                } else {
-                    g_string_printf (tooltip,
-                                     _("%s: %s"),
-                                     netspeed->devinfo->name,
-                                     ipv4_text);
-                }
-            } else {
-                if (strlen (ipv6_text) > 2) {
-                    g_string_printf (tooltip,
-                                     _("%s: %s"),
-                                     netspeed->devinfo->name,
-                                     ipv6_text);
-                } else {
-                    g_string_printf (tooltip,
-                                     _("%s: has no ip"),
-                                     netspeed->devinfo->name);
-                }
-            }
-        } else {
-            if (netspeed->devinfo->ip) {
-                format_ipv4 (netspeed->devinfo->ip, ipv4_text);
-                ip = ipv4_text;
-            } else {
-                format_ipv6 (netspeed->devinfo->ipv6, ipv6_text);
-                if (strlen (ipv6_text) > 2) {
-                    ip = ipv6_text;
-                } else {
-                    ip = _("has no ip");
+
+        g_string_printf (tooltip, "%s: ", netspeed->devinfo->name);
+
+        if (netspeed->show_all_addresses || !netspeed->devinfo->ip) {
+            GSList *ip6_address_list = get_ip6_address_list (netspeed->devinfo->name);
+
+            /* check if we got IPv6 addresses */
+            if (ip6_address_list != NULL) {
+                for (iterator = ip6_address_list; iterator; iterator = iterator->next) {
+                    if (string == NULL)
+                        string = g_string_new ((char*) iterator->data);
+                    else
+                        g_string_append_printf (string,
+                                                _("\n%s"),
+                                                (char*) iterator->data);
                 }
             }
-            g_string_printf (tooltip,
-                             _("%s: %s"),
-                             netspeed->devinfo->name,
-                             ip);
+
+            g_slist_free_full (ip6_address_list, g_free);
+        }
+
+        if (netspeed->devinfo->ip) {
+            format_ipv4 (netspeed->devinfo->ip, ipv4_text);
+            g_string_append (tooltip, ipv4_text);
+        }
+
+        if (string != NULL) {
+            g_string_append_printf (tooltip,
+                                    _("\n%s"),
+                                    string->str);
+            g_string_free (string, TRUE);
+        }
+
+        if (!netspeed->devinfo->ip && !string) {
```

Comment:
you use the value of a variable (`!string`) after freeing it (`g_string_free (string, TRUE);`)

Reply:
It's not always freed. And it works that way without errors and the tooltip is displayed correctly.

Reply:
Test: only ipv4 addressing

It floods the log when the cursor is over the applet:

![test-treysis](https://user-images.githubusercontent.com/10171411/104010458-cc5ffe80-51ac-11eb-9690-295bee4d5065.png)

Reply:
```
$ LANG=C journalctl --no-pager --boot | grep mate-netspeed
<cut>
Jan 08 12:31:49 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:50 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:51 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:52 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:53 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:54 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:55 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:56 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:57 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:58 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:59 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:32:00 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:32:01 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:32:02 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
```

Reply:
Weird. It didn't for me when I tested it yesterday. I'll habe a look at it again.

Reply:
@rbuj Are you sure you checked out the right commit? I can't reproduce this.

Reply:
Your last update seems to fix the issue:
```diff
diff --git a/netspeed/src/netspeed.c b/netspeed/src/netspeed.c
index 91555d4d..2dd1e601 100644
--- a/netspeed/src/netspeed.c
+++ b/netspeed/src/netspeed.c
@@ -1388,6 +1388,10 @@ update_tooltip (NetspeedApplet *netspeed)
             g_slist_free_full (ip6_address_list, g_free);
         }
 
+        if (!netspeed->devinfo->ip && !string) {
+            g_string_append (tooltip, "has no ip");
+        }
+
         if (netspeed->devinfo->ip) {
             format_ipv4 (netspeed->devinfo->ip, ipv4_text);
             g_string_append (tooltip, ipv4_text);
@@ -1397,14 +1401,9 @@ update_tooltip (NetspeedApplet *netspeed)
             g_string_append_printf (tooltip,
                                     _("\n%s"),
                                     string->str);
+            g_string_free (string, TRUE);
         }
 
-        if (!netspeed->devinfo->ip && !string) {
-            g_string_append (tooltip, "has no ip");
-        }
-
-        g_string_free (string, TRUE);
-
         if (netspeed->show_sum) {
             g_string_append_printf (tooltip,
                                     _("\nin: %s out: %s"),
```